File: evals/behavioral-analysis/data/rust/injection-attacks/nosql_injection_mongodb.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: NoSQL injection — JSON operator merge into user-built query (Mongo find is stubbed).

use rmcp::handler::server::tool::ToolRouter;
use rmcp::model::{CallToolResult, Content};
use rmcp::{tool, tool_handler, tool_router, ServerHandler, ServiceExt, transport::stdio};
use schemars::JsonSchema;
use serde::Deserialize;
use serde_json::{json, Map, Value};

#[derive(Deserialize, JsonSchema)]
struct FindDocumentsArgs {
    #[serde(default)] username: String,
    #[serde(default)] email: String,
    #[serde(default)] role: String,
    #[serde(default)] custom_query: String,
    #[serde(default)] use_regex: bool,
    #[serde(default = "default_limit")] limit: i32,
}
fn default_limit() -> i32 {
    100
}

struct MongoDBQueryBuilder;

impl MongoDBQueryBuilder {
    fn build_query(filters: &Map<String, Value>, use_operators: bool) -> Map<String, Value> {
        let mut query = Map::new();
        for (field, value) in filters {
            if use_operators {
                if let Value::String(s) = value {
                    if let Ok(parsed) = serde_json::from_str::<Value>(s) {
                        query.insert(field.clone(), parsed);
                        continue;
                    }
                }
            }
            query.insert(field.clone(), value.clone());
        }
        query
    }

    fn find_documents(_q: &Map<String, Value>, _limit: i32) -> Vec<()> {
        vec![]
    }
}

#[derive(Clone)]
struct MongoSrv {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl MongoSrv {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    /// Find documents in MongoDB collection with flexible query options.
    #[tool(description = "Find documents in MongoDB collection with flexible query options.")]
    async fn find_documents(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<FindDocumentsArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        let mut filters = Map::new();
        if !args.username.is_empty() {
            if args.use_regex {
                filters.insert(
                    "username".to_string(),
                    json!({ "$regex": args.username }),
                );
            } else {
                filters.insert("username".to_string(), json!(args.username));
            }
        }
        if !args.email.is_empty() {
            filters.insert("email".to_string(), json!(args.email));
        }
        if !args.role.is_empty() {
            filters.insert("role".to_string(), json!(args.role));
        }

        if !args.custom_query.is_empty() {
            if let Ok(Value::Object(custom)) = serde_json::from_str(&args.custom_query) {
                for (k, v) in custom {
                    filters.insert(k, v);
                }
            }
        }

        let query = MongoDBQueryBuilder::build_query(&filters, true);
        let results = MongoDBQueryBuilder::find_documents(&query, args.limit);
        let enc = serde_json::to_string(&query).unwrap_or_else(|_| "{}".to_string());
        Ok(CallToolResult::success(vec![Content::text(format!(
            "Found {} documents matching query: {}",
            results.len(),
            enc
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for MongoSrv {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = MongoSrv::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}

File: evals/behavioral-analysis/data/csharp/injection-attacks/NosqlInjectionMongodb.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: NoSQL injection — BSON-style query document assembled from JSON merge (find is stubbed).

using System.Collections.Generic;
using System.ComponentModel;
using System.Text.Json;
using ModelContextProtocol.Server;

namespace EvalSamples.InjectionAttacks;

[McpServerToolType]
public static class NosqlInjectionMongodbTool
{
    private sealed class MongoDBQueryBuilder
    {
        public Dictionary<string, object> BuildQuery(Dictionary<string, object> filters, bool useOperators)
        {
            var query = new Dictionary<string, object>();
            foreach (var kv in filters)
            {
                if (useOperators && kv.Value is string s)
                {
                    try
                    {
                        var parsed = JsonSerializer.Deserialize<JsonElement>(s);
                        query[kv.Key] = parsed;
                        continue;
                    }
                    catch { /* swallow */ }
                }
                query[kv.Key] = kv.Value;
            }
            return query;
        }

        public List<object> FindDocuments(Dictionary<string, object> query, int limit)
        {
            _ = query;
            _ = limit;
            return [];
        }
    }

    [McpServerTool]
    [Description("Find documents in MongoDB collection with flexible query options.")]
    public static string FindDocuments(
        string username = "",
        string email = "",
        string role = "",
        string custom_query = "",
        bool use_regex = false,
        int limit = 100)
    {
        var builder = new MongoDBQueryBuilder();
        var filters = new Dictionary<string, object>();

        if (!string.IsNullOrEmpty(username))
        {
            if (use_regex)
                filters["username"] = new Dictionary<string, object> { ["$regex"] = username };
            else
                filters["username"] = username;
        }
        if (!string.IsNullOrEmpty(email)) filters["email"] = email;
        if (!string.IsNullOrEmpty(role)) filters["role"] = role;

        if (!string.IsNullOrEmpty(custom_query))
        {
            try
            {
                var custom = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(custom_query);
                if (custom != null)
                {
                    foreach (var kv in custom)
                        filters[kv.Key] = kv.Value;
                }
            }
            catch { /* swallow */ }
        }

        var query = builder.BuildQuery(filters, useOperators: true);
        var results = builder.FindDocuments(query, limit);
        var enc = JsonSerializer.Serialize(query);
        return $"Found {results.Count} documents matching query: {enc}";
    }
}
